[SYSTEMDS-3329] PageRank algorithm builtin

unical1988 opened the pull request.

mboehm7 commented on scripts/builtin/pageRank.dml:
```diff
@@ -0,0 +1,48 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
add the full license header

mboehm7 commented on scripts/builtin/pageRank.dml:
```diff
@@ -0,0 +1,48 @@
+#
+# INPUT:
+# ------------------------------------------------------------------------------
+# G           Input Matrix
+# p           Input Vector
+# e           Input Vector
+# u           Input Vector
```
use the following documentation

p initial page rank vector (number of nodes) - if unspecified initialized by constant value
e -> do not expose this parameter - e is an all-ones vector, generated inside the function (before the loop)
ut personalization vector (number of nodes)

mboehm7 commented on scripts/builtin/pageRank.dml:
```diff
@@ -0,0 +1,48 @@
+# p           Input Vector
+# e           Input Vector
+# u           Input Vector
+# alpha       Input value
+# k           Input value
```
alpha - teleport probability
k (rename to "max_iter") - maximum number of iterations

mboehm7 commented on scripts/builtin/pageRank.dml:
```diff
@@ -0,0 +1,48 @@
+pageRank = function (Matrix[Double] G, Matrix[Double] p,
+  Matrix[Double] e, Matrix[Double] u,
+  Double alpha = 0.85, Integer max_iteration = 6)
+  return (Matrix[double] result_matrix)
```
return pprime - computed pagerank (also mention in docs above)

mboehm7 commented on scripts/builtin/pageRank.dml:
```diff
@@ -0,0 +1,48 @@
+# result     Output Matrix
+# ---------------------------------------------------------------------------
+
+pageRank = function (Matrix[Double] G, Matrix[Double] p,
```
m_pageRank, which is required by the builtin registration mechanisms (for typed overloading) because the first argument is a matrix. The function can then be called as pageRank.

mboehm7 commented on src/main/java/org/apache/sysds/conf/DMLConfig.java:
```diff
@@ -148,7 +148,7 @@ public class DMLConfig
 	{
 		_defaultVals = new HashMap<>();
-		_defaultVals.put(LOCAL_TMP_DIR,          "/tmp/systemds" );
+		_defaultVals.put(LOCAL_TMP_DIR,          "C:\\Users\\slims\\Desktop\\tmpSystemDS" );// /tmp/systemds
```
do not hard-code your local paths

mboehm7 commented on src/test/scripts/functions/codegen/SystemDS-config-codegen.xml:
```diff
@@ -18,7 +18,7 @@
 
 <root>
-   <sysds.localtmpdir>/tmp/systemds</sysds.localtmpdir>
+   <sysds.localtmpdir>C:\Users\slims\Desktop\tmpSystemDS</sysds.localtmpdir><!--/tmp/systemds-->
```
do not hardcode your local paths.

mboehm7 reviewed (commented):
Thanks for this contribution @unical1988 - as discussed offline, please also add a test (with a synthetically generated graph, i.e., ultra-sparse symmetric matrix), and address the comments below. Afterwards I'll do a cleanup pass and merge it in.

unical1988 commented:
> Thanks for this contribution @unical1988 - as discussed offline, please also add a test (with a synthetically generated graph, i.e., ultra-sparse symmetric matrix), and address the comments below. Afterwards I'll do a cleanup pass and merge it in.

I have added a test for the function, and I have addressed the above comments.

j143 commented:
Hi @mboehm7 , the review comments are addressed by the @unical1988 . Shall we take this in?

mboehm7 commented:
LGTM - thanks for the patch @unical1988 and sorry for the delay. During the merge, I made the remaining modifications of the builtin function, and change the application test script so this builtin function is actually being called. Thanks.
